framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,656 @@
+        int componentEndPos;
+        {
+            componentEndPos = arrayType.indexOf(' ');
+            if (arrayType.startsWith("@")) {
+                // While there are more explicit annotations, go to the next space.
+                while (arrayType.charAt(componentEndPos + 1) == '@') {
+                    componentEndPos = arrayType.indexOf(' ', componentEndPos + 1);
+                }
+                // Once all annotations have been skipped, we still have to skip to the end of the
+                // component type itself.
+                componentEndPos = arrayType.indexOf(' ', componentEndPos + 1);
+            }
+            // If the first array level doesn't have an annotation, then the spacing will be off,
+            // so use the position of the first '[' instead (which must be correct, since the
+            // first array doesn't have an annotation).
+            int firstArrayPos = arrayType.indexOf('[');
+            if (firstArrayPos < componentEndPos || componentEndPos == -1) {
+                componentEndPos = firstArrayPos - 1;
+            }
+        }
+
+        String componentType = arrayType.substring(0, componentEndPos + 1);
+        String arrayTypes = arrayType.substring(componentEndPos + 1);
+
+        return formatArrayTypeImpl(e, arrayTypes, componentType, result) + " ";
+    }
+
+    /**
+     * The implementation of formatArrayType. Java array types have a somewhat unintuitive syntax:
+     * see <a
+     * href="https://checkerframework.org/jsr308/specification/java-annotation-design.html#array-syntax">this
+     * explanation</a>.
+     *
+     * <p>The iteration order in the scene-lib representation is from outermost to innermost. For
+     * example, given the type {@code @Foo int @Bar [] @Baz []}, the iteration order is {@code @Bar
+     * []}, then {@code @Baz []}, and then finally {@code @Foo int}. This implementation therefore
+     * passes a string builder with the result of the array types seen so far, to handle
+     * multidimensional arrays. That builder is appended to the final component type in the base
+     * case.
+     *
+     * @param e same as above, but can become null if scene-lib did not fill in the inner types,
+     *     which happens when they do not have annotations
+     * @param arrayTypes the array parts of the array type (i.e. the parts after the component type
+     *     in the string representation). Must contain at least one '['.
+     * @param componentType the component type of the array
+     * @param result the string builder containing the array types seen so far
+     * @return the formatted string, without a trailing space
+     */
+    private static String formatArrayTypeImpl(
+            @Nullable ATypeElement e,
+            String arrayTypes,
+            String componentType,
+            StringBuilder result) {
+        // append the next type:
+        String nextArrayType = arrayTypes.substring(0, arrayTypes.indexOf(']') + 1);
```

Comment:
What if an annotation's printed representation includes "]"?  Maybe that doesn't happen because `arrayTypes` is guaranteed to have no annotations?